### PR TITLE
feat(WhiteBox): implement transform mode

### PR DESCRIPTION
**Pull request by pollend**
This adds the ability to do fine tuning adjustment in whitebox. Early work in progress on this feature. I'm not quite sure how this will work functionally.

### CODE REVIEW

**Comment by pollend** [comment]
@hultonha here is a wip for this what do you think about this functionally?

https://user-images.githubusercontent.com/854359/166184380-38803fea-07d0-40a0-998f-38a170709f2e.mp4

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:88]: suggestion: In future we might want to actually add a new button for `SubMode::Transform` instead of hiding them, I'd recommend we move to this if we can

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:497]: uber nit: Please remove the extra line break

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:88]: (sorry kept reading and I think I see what is happening sorry 😅 this actually could be okay (I'll check out the video soon too)

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:520]: nit: Update `select_object` to `TransformMode` potentially?

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:81]: uber nit: Missing space after `//!`

File: Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:22]: uber nit: Brace placement (braces always on a new line)

File: Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:28]: uber nit: need line break here (between function definitions)

File: Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:42]: uber nit: need EOL character

File: Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.h
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.h:36]: nit: closing namespace and EOL

**Review by hultonha — COMMENTED** [review]
This looks like an awesome stub @pollend! This is exactly what I meant on the call, great work putting it together! 😄 Might be worth having a look at `EditorTransformComponentSelection.cpp` for some examples of how we use Manipulators (and also all the Modifiers that exist in the White Box Gem). But you probably want something similar to what's happening for the normal entity manipulation... but the good news is a lot of the new code can go into this new mode you added and not impact any of the other modes 🙂

Fantastic first start!

(we might want to even consider tidying up this stub, and then disabling it by default, so you could merge this change in with it 'off' - before doing lots more work, this will help us make incremental changes and iterate as we go, instead of having a MASSIVE PR we need to review in future. We could use a settings registry value for this and then have some `if (newModeEnabled)` checks in a couple of places.... Let me know what you think, but we're all about incremental dev now which will make this easier I reckon

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (pollend) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:520]: umm assets ?

Comment (hultonha) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp]: Aaah got you, sorry is this referring to an existing asset for the icon? Hmm we can try and ask UX to get us a new one, in the meantime this is okay to stay as-is then, sorry for the confusion

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:26]: update: Please group this with other `AzCore` includes

Comment (hultonha) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:26]: (and leave a line break between includes and the opening namespace)

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:29]: uber nit: Can this just be on one line (line limit is 140 characters and this is only 107 when all on one line

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:534]: question: By doing this, when the setting is 'off/disabled' is it impossible to get into this mode? We just want to make sure this feature is 'invisible' to users unless they have the setting on. It would be good to do some testing to make sure nothing blows up when it's disabled.

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:103]: nit: Please remove extra line break (we aim for maximum one)

**Review by hultonha — COMMENTED** [review]
Updates are looking good, I have a few more comments and I've added the UX team for visibility to check out the video and give feedback. I really like the idea but want to make sure the UX sig is involved to offer direction. Thanks again, great progress so far! 🙂

**Comment by pollend** [comment]
@hultonha I verified this with editorpreferences.setreg. so the new button on the bottom of the list is not there because we don't have an icon for it. where would that come from?

**Comment by hultonha** [comment]
> @hultonha I verified this with editorpreferences.setreg. so the new button on the bottom of the list is not there because we don't have an icon for it. where would that come from?

That's something we'll need UX help from - I think reusing an existing one in the short-term is okay, but perhaps @amzn-leenguy may be able to help here and attach one to this PR that you could include. The icon should represent advanced 'transform' editing if that's any help Lee. Thanks!

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:530]: nit: Please always have braces start on a new line - I'd recommend using `clang-format` to do this, but I think White Box used a slightly different style (120 character limit instead of 140). So may run a formatting pass as a separate formatting only change to make things simpler in future.

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:83]: uber nit: Might be nice to group all the `Enter` functions together so I'd lose the line break on line 80

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:110]: I've format these normally so they're all on the same line (and if they need a comment use `//!` on the line above as opposed to `//!<` on the formatting for line length looks a bit strange)

**Review by hultonha — APPROVED** [review]
This looks like a great start to me! I am happy to approve this and we can then iterate on it in future 👍 

For reference you should be able to toggle the setreg value from the console like this to test it more quickly

```
sr_regset //O3DE/Preferences/WhiteBox/TransformFeature true
// or
sr_regset //O3DE/Preferences/WhiteBox/TransformFeature false
```

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (pollend) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:530]: Need to remember to run clang format

**Comment by hultonha** [comment]
Looks like there were some failures here - https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-9331/

**Comment by pollend** [comment]
> Looks like there were some failures here - https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-9331/

I was wondering why EdgeRestoreMode had a ton of forward declarations. why would this be the case, I don't quite follow

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (hultonha) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp]: Just a quick FYI be careful with this as you may get more changes than you expect, so it might be worth doing this as a separate pass, just something to be aware of

Comment (hultonha) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp]: Ah nice! That's a great idea 🙂 Well remembered 👍

**Review by AMZN-daimini — COMMENTED** [review]
This is great :D A few issues to fix, but nothing major.

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.h:120]: update: If you're removing the `<`, please move the comment above the variable declaration

e.g.

```
//! Event handler for sub mode changes.
AZ::Event<AzToolsFramework::ViewportUi::ButtonId>::Handler m_modeSelectionHandler;
```

If you do this can you move everything back on to its own line as above, thanks!

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentModeTypes.h
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentModeTypes.h:24]: uber nit: Please remove the additional line break, in the White Box Gem we aim for 1 max line breaks between code segments, no more

File: Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp
Comment (hultonha) [reply Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp:1]: +1 good spot!

**Review by hultonha — COMMENTED** [review]
Good spot by @AMZN-daimini on the copyright headers, once they're added I'll approve again, I have a couple of other minor nits too

File: Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp
Comment (pollend) [reply Gems/WhiteBox/Code/Source/EditorWhiteBoxComponentMode.cpp:540]: I don't think you can m_defaultModeButtonId is not a constant

**Review by hultonha — APPROVED** [review]

**Comment by cgalvan** [comment]
@pollend  There was a compile error on the AR run:

```
[2022-05-24T18:49:13.841Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(9,10): fatal error C1083: Cannot open include file: 'EditorWhiteBoxTransformMode.h': No such file or directory [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]
[2022-05-24T18:49:13.841Z]   #include <EditorWhiteBoxTransformMode.h>
```
Might be a unity vs. non-unity build issue?

**Comment by pollend** [comment]
@cgalvan discovered clangd really nice to see that it highlighted the mistake 

image

**Comment by hultonha** [comment]
Just kicked off another build, thanks @pollend 👍

**Comment by cgalvan** [comment]
> @cgalvan discovered clangd recently; really nice to see that it highlighted the mistake
> 
> image

Nice I didn't know about that :)

**Comment by cgalvan** [comment]
@pollend  Looks like a few more compile issues cropped up:

```
[2022-05-26T09:45:30.156Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(31,117): error C2220: the following warning is treated as an error [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-05-26T09:45:30.165Z]       AZStd::vector<AzToolsFramework::ActionOverride> TransformMode::PopulateActions(const AZ::EntityComponentIdPair& entityComponentIdPair)

[2022-05-26T09:45:30.165Z]                                                                                                                       ^

[2022-05-26T09:45:30.422Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(31,117): warning C4100: 'entityComponentIdPair': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-05-26T09:45:30.431Z]       AZStd::vector<AzToolsFramework::ActionOverride> TransformMode::PopulateActions(const AZ::EntityComponentIdPair& entityComponentIdPair)

[2022-05-26T09:45:30.718Z]                                                                                                                       ^

[2022-05-26T09:45:30.986Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(41,44): warning C4100: 'debugDisplay': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-05-26T09:45:31.253Z]           AzFramework::DebugDisplayRequests& debugDisplay)

[2022-05-26T09:45:31.510Z]                                              ^

[2022-05-26T09:45:31.520Z]      Creating library D:/workspace/o3de/build/windows/lib/profile/GradientSignal.lib and object D:/workspace/o3de/build/windows/lib/profile/GradientSignal.exp

[2022-05-26T09:45:31.520Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(40,42): warning C4100: 'viewportInfo': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-05-26T09:45:31.520Z]           const AzFramework::ViewportInfo& viewportInfo,

[2022-05-26T09:45:31.778Z]                                            ^

[2022-05-26T09:45:32.343Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(39,42): warning C4100: 'renderData': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-05-26T09:45:32.352Z]           const IntersectionAndRenderData& renderData,

[2022-05-26T09:45:37.496Z]                                            ^

[2022-05-26T09:45:37.504Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(38,30): warning C4100: 'worldFromLocal': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-05-26T09:45:37.504Z]           const AZ::Transform& worldFromLocal,

[2022-05-26T09:45:37.504Z]                                ^

[2022-05-26T09:45:37.504Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(37,42): warning C4100: 'entityComponentIdPair': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]

[2022-05-26T09:45:37.504Z]   unity_0_cxx.cxx

[2022-05-26T09:45:37.504Z]           const AZ::EntityComponentIdPair& entityComponentIdPair,

[2022-05-26T09:45:37.504Z]   unity_7_cxx.cxx

[2022-05-26T09:45:37.504Z]                                            ^

[2022-05-26T09:45:37.505Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(50,52): warning C4100: 'vertexIntersection': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]
[2022-05-26T09:45:37.505Z]           const AZStd::optional<VertexIntersection>& vertexIntersection)
[2022-05-26T09:45:37.505Z]                                                      ^
[2022-05-26T09:45:37.505Z]   unity_4_cxx.cxx
[2022-05-26T09:45:37.505Z] :/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(49,53): warning 4100: 'polygonIntersection': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-85ae382\Code\WhiteBox.Editor.Static.vcxproj]
[2022-05-26T09:45:37.505Z]           const AZStd::optional<PolygonIntersection>& polygonIntersection,
[2022-05-26T09:45:37.505Z]                                                       ^
[2022-05-26T09:45:37.505Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(48,50): warning C4100: 'edgeIntersection': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]
[2022-05-26T09:45:37.505Z]           const AZStd::optional<EdgeIntersection>& edgeIntersection,
[2022-05-26T09:45:37.505Z]                                                    ^
2022-05-26T09:45:37.505Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(47,42): warning C4100: 'entityComponentIdPair': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]
[2022-05-26T09:45:37.505Z]           const AZ::EntityComponentIdPair& entityComponentIdPair,
[2022-05-26T09:45:37.505Z]                                            ^
[2022-05-26T09:45:37.505Z] D:/workspace/o3de/Gems/WhiteBox/Code/Source/SubComponentModes/EditorWhiteBoxTransformMode.cpp(46,77): warning C4100: 'mouseInteraction': unreferenced formal parameter [D:\workspace\o3de\build\windows\External\WhiteBox-285ae382\Code\WhiteBox.Editor.Static.vcxproj]
[2022-05-26T09:45:38.060Z]   unity_8_cxx.cxx
[2022-05-26T09:45:38.069Z]           const AzToolsFramework::ViewportInteraction::MouseInteractionEvent& mouseInteraction,
2022-05-26T09:45:38.069Z]                                                                               ^
[2022-05-26T09:45:41.413Z]   ScriptCanvasEditorGem.cpp
```

**Comment by hultonha** [comment]
Ah good spot @cgalvan, might need a couple of `[[maybe_unused]]` attributes to be added @pollend. Thanks!

**Comment by amzn-leenguy** [comment]
Hi Team, 

As UX representative on this feature, I wont be inputting on the technical code part of this yet. I will be more interested in the final UX of the update. Please upload a video when you require me to review anything for approval.

Thanks! CC @hultonha @pollend

**Comment by pollend** [comment]
> Hi Team,
> 
> As UX representative on this feature, I wont be providing any feedback on the technical code part of this yet. I will be more interested in the final UX of the update. Please upload a video when you require me to review anything for approval.
> 
> Thanks! CC @hultonha @pollend

https://user-images.githubusercontent.com/854359/166184380-38803fea-07d0-40a0-998f-38a170709f2e.mp4

**Comment by hultonha** [comment]
Thanks for updating @pollend, I'll kick off another build now 👍

**Comment by pollend** [comment]
> Thanks for updating @pollend, I'll kick off another build now +1

that CI error seems unrelated

**Comment by cgalvan** [comment]
@pollend Can you do another conflict resolve? I think it got confused since I merged in your other fix with the tooltips.

**Comment by hultonha** [comment]
@pollend Could you please merge development into this branch again (or rebase), I think this failing test with large worlds was fixed recently (tagging @mbalfour-amzn for visibility) 

Failure was  **Failed: Test test_Terrain_SupportsPhysics:**

https://jenkins.build.o3de.org/blue/organizations/jenkins/O3DE/detail/PR-9331/8/pipeline/617

**Review by cgalvan — APPROVED** [review]

**Comment by pollend** [comment]
@cgalvan one thing to note is an icon was never added for this new mode. did we ever get a response back from @amzn-leenguy

**Comment by hultonha** [comment]
> @cgalvan one thing to note is an icon was never added for this new mode. did we ever get a response back from @amzn-leenguy

As this mode is behind a CVAR this isn't the end of the world, but we get one before we enable it for real. Unfortunately @amzn-leenguy is going to be out for a little while, perhaps @bhanuja-s might be able to help us with an icon? 😁

**Pull request merged**
